designs/2023-test-rule-errors/README.md

```diff
@@ -0,0 +1,295 @@
+            throw err;
+        }
+    }
+    ```
+
+I'm interested to gather feedback on these approaches and preferences about what the implementation should look like.
+
+## Documentation
+
+<!--
+    How will this RFC be documented? Does it need a formal announcement
+    on the ESLint blog to explain the motivation?
+-->
+
+We will document this change in the ESLint [RuleTester](https://eslint.org/docs/latest/developer-guide/nodejs-api#ruletester) section of the [Node.JS API](https://eslint.org/docs/latest/developer-guide/nodejs-api) page.
+
+Ideally, we can also draw attention to it with a paragraph and example usage in the blog post / release notes for the version it ships in. This is the least we can do to raise awareness of it for rule authors who may be interested to take advantage of it in new or existing rules.
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think
+    about any opposing viewpoints that reviewers might bring up.
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+- A user testing that their rule schema catches invalid options may end up including the message from JSON Schema in their test case (e.g. `Value "bar" should be boolean.`). This could make it more difficult for ESLint to upgrade its version of ajv / JSON Schema in the future, as an upgraded version could tweak messages and necessitate updating any test cases that include the message. This is a relatively minor concern, as it's unlikely that messages will change often, and it's easy to update test cases if they do. Two factors that could mitigate this:
+  - If [snapshot testing](https://github.com/eslint/eslint/issues/14936) is implemented in the future, it would become possible to fix test cases with a single command.
+  - We can encourage the use of `messageId` (discussed in [Detailed Design](#detailed-design)) for the majority of cases that don't care about including the exact message text.
+
+## Backwards Compatibility Analysis
+
+<!--
+    How does this change affect existing ESLint users? Will any behavior
+    change for them? If so, how are you going to minimize the disruption
+    to existing users?
+-->
+
+This new feature is a non-breaking change and has no impact unless a user chooses to use it.
+
+## Alternatives
+
+<!--
+    What other designs did you consider? Why did you decide against those?
+
+    This section should also include prior art, such as whether similar
+    projects have already implemented a similar feature.
+-->
+
+### Top-level `error` array
+
+Instead of using the `invalid` test cases to cover this new functionality, we could have a dedicated `error` array of test cases for testing exceptions:
```

Comment:
I'm more in favor of this design. `invalid` is for successful executions of the rule and testing whether it provides expected lint errors. I think it would be confusing that array entries (test cases) with `fatal: true` have similar but different shapes, constraints, and especially different meanings for same properties (`message` would relate to exception messages, not lint messages; `messageId` would be an enum, not a lint message id).

Reply:
I'm open to switching to the `error` array design and I agree with your reasoning. Let's see what others think. The original suggestion to re-use `invalid` came from [here](https://github.com/eslint/eslint/issues/13434#issuecomment-648215748).

It could also be confusing or disorganized to have fatal error test cases scattered throughout non-fatal error test cases in the same array.

Reply:
I also favor another array, though I might choose `throws` to be extra clear about what is being tested. I think `error` is too close to `invalid` and may be a bit confusing.

In any event, I think the reasoning for having a different array is sound. We are no longer testing the validity of the code, we are now testing the robustness of the rule itself should it encounter unexpected input or conditions. And in this case, we could allow `code` to be optional.

Reply:
I'll update the RFC to the separate array. 

Any preference between `exception` and `throws`? I prefer `exception` since the valid/invalid aren't verbs.

Reply:
I think the name `throws` best describes what is being tested.

Reply:
Can you elaborate? `exception` and `throws` both refer to the same thing. How about `fatal`? It's short and more consistent with `valid` / `invalid` while not overly-specific.

Reply:
I have updated the RFC design to be based on a separate `fatal` array.